// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";

package optimization_guide.proto;

import "components/optimization_guide/proto/features/common_quality_data.proto";

option java_outer_classname = "ActionsProto";

option java_package = "org.chromium.components.optimization_guide.features.proto";

option optimize_for = LITE_RUNTIME;

// DO NOT EDIT THIS FILE DIRECTLY!
//
// This file is generated in g3 and then synced to Chrome. Instead, please
// refer to http://go/chrome-intelligence-feature-protos (Google-internal link),
// and then changes will be synced with Chrome automatically.

// The target of an action.
// Next ID: 4
message ActionTarget {
  // Id of the content node within the frame. Same as
  // ContentNode.content_node_id.
  int32 content_node_id = 1;

  Coordinate coordinate = 2;

  // Which document in a frame the action should be taken in.
  DocumentIdentifier document_identifier = 3;
}

// Mouse click action.
// Next ID: 4
message ClickAction {
  ActionTarget target = 1;

  ClickType click_type = 2;

  ClickCount click_count = 3;

  enum ClickType {
    // Unknown click type. This should not be used.
    UNKNOWN_CLICK_TYPE = 0;

    // Left mouse click.
    LEFT = 1;

    // Right mouse click.
    RIGHT = 2;
  }

  // Represents the number of times the target was clicked, for now only
  // supports single and double clicks.
  enum ClickCount {
    // Unknown click count. This should not be used.
    UNKNOWN_CLICK_COUNT = 0;

    // Single click.
    SINGLE = 1;

    // Double click.
    DOUBLE = 2;
  }
}

// Type text action.
// Next ID: 5
message TypeAction {
  ActionTarget target = 1;

  // The text to type.
  string text = 2;

  // After Typing the text, press enter.
  bool follow_by_enter = 3;

  TypeMode mode = 4;

  // How the text should be inserted into the target.
  enum TypeMode {
    UNKNOWN_TYPE_MODE = 0;

    DELETE_EXISTING = 1;

    PREPEND = 2;

    APPEND = 3;
  }
}

// Scroll action.
// Next ID: 4
message ScrollAction {
  ActionTarget target = 1;

  ScrollDirection direction = 2;

  // The distance to scroll in pixels.
  float distance = 3;

  enum ScrollDirection {
    // Unknown scroll direction. This should not be used.
    UNKNOWN_SCROLL_DIRECTION = 0;

    // Scroll left.
    LEFT = 1;

    // Scroll right.
    RIGHT = 2;

    // Scroll up.
    UP = 3;

    // Scroll down.
    DOWN = 4;
  }
}

// Move mouse action.
// Next ID: 2
message MoveMouseAction {
  ActionTarget target = 1;
}

// Drag and release mouse action.
// Next ID: 3
message DragAndReleaseAction {
  ActionTarget from_target = 1;

  ActionTarget to_target = 2;
}

// Select dropdown action.
// Next ID: 3
message SelectAction {
  ActionTarget target = 1;

  // The value of the option to select.
  string value = 2;
}

// Navigate the page to a URL action.
// Next ID: 2
message NavigateAction {
  string url = 1;
}

// Wait action waits 3 seconds by default.
// Next ID: 1
message WaitAction {
}

// Navigate the page Back action.
// Next ID: 1
message HistoryBackAction {
}

// Navigate the page Forward action.
// Next ID: 1
message HistoryForwardAction {
}

// All of the actions that can be taken.
// Only one of these is allowed per action.
// Next ID: 11
message ActionInformation {
  oneof action_info {
    ClickAction click = 1;

    TypeAction type = 2;

    ScrollAction scroll = 3;

    MoveMouseAction move_mouse = 4;

    DragAndReleaseAction drag_and_release = 5;

    SelectAction select = 6;

    NavigateAction navigate = 7;

    HistoryBackAction back = 8;

    HistoryForwardAction forward = 9;

    WaitAction wait = 10;
  }
}

// A single set of actions to be taken by Chrome before responding.
// Next ID: 2
message BrowserAction {
  // The actions to be taken by Chrome.
  repeated ActionInformation action_information = 1;
}
